js/TableDetailComponent.js

```diff
@@ -0,0 +1,97 @@
+class TableDetailComponent {
+    
+    /**
+     * TableDetailComponent
+     * @constructor
+     * @param {jQuery} $html - jQuery wrapper of the html node
+     */
+    constructor ($html) {
+        this.$html = $html;
+    }
+
+    /**
+     * Initialise
+     */
+    init () {
+        if (typeof this.$html === 'undefined' || !this.$html) {
+            throw new Error('$html must be passed to TableDetailComponent');
+        }
+
+        let $panelHtml = $(
+            '<div class="table-detail t-table-detail" data-table-detail-panel>' +
+            '   <div class="table-detail__header">' +
+            '       <button type="button" class="close table-detail__header-close" data-table-detail-close-panel aria-hidden="true">&times;</button>' +
+            '       <h4 class="table-detail__title" data-table-detail-panel-title>Detail</h4>' +
+            '   </div>' +
+            '   <div class="table-detail__body" data-table-detail-panel-body></div>' +
+            '</div>'
+        );
+
+        this.$table = this.$html.find('[data-table-detail-table]');
+
+        // Add backdrop and detail panel if UI contains a table detail pattern
+        if (this.$table) {
+            this.$html.find('body')
+                .append('<div class="table-detail-backdrop"></div>')
+                .append($panelHtml);
+        }
+
+        // Grap detail panel bits we need
+        this.$detailPanel = this.$html.find('[data-table-detail-panel]');
+        this.$detailPanelBody = this.$html.find('[data-table-detail-panel-body]');
+        this.$detailPanelTitle = this.$html.find('[data-table-detail-panel-title]');
+
+        // Open click listener
+        this.$table.find('[data-table-detail-view-detail]').on('click', (event) => {
+            event.preventDefault();
+
+            let detailContent = $(event.currentTarget).closest('tr').data('table-detail-content');
+            let customDetailPanelTitle = $(event.currentTarget).closest('tr').data('table-detail-panel-custom-title');
+            
+            this.viewDetail(detailContent, customDetailPanelTitle);
+        });
+
+        // Close click listener
+        this.$detailPanel.find('[data-table-detail-close-panel]').on('click', (event) => {
```

Comment:
Super minor comment, which is open for opinion really. We _could_ omit anonymous arrow function brackets if it accepts only 1 argument (although I'm not sure what we really gain from that!) – also worth noting that we cannot do that in most cases when writing TypeScript, thought I'd mention it though:

```
node.addEventListener('click', event => {
    event.preventDefault();
});
```